pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -482,3 +455,90 @@ func createCommandForExecutingRunScript(scriptsURL map[string]string, location s
 	}
 	return cmd
 }
+
+func downloadAndExtractFileFromContainer(docker dockerpkg.Docker, tar s2itar.Tar, sourcePath, destinationPath, containerID string) (api.FailureReason, error) {
+	glog.V(5).Infof("Downloading file %q", sourcePath)
+
+	fd, err := ioutil.TempFile(destinationPath, "s2i-runtime-artifact")
+	if err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonFSOperationFailed,
+			utilstatus.ReasonMessageFSOperationFailed,
+		)
+		return res, fmt.Errorf("could not create temporary file for runtime artifact: %v", err)
+	}
+	defer func() {
+		fd.Close()
+		os.Remove(fd.Name())
+	}()
+
+	if err := docker.DownloadFromContainer(sourcePath, fd, containerID); err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return res, fmt.Errorf("could not download file (%q -> %q) from container %s: %v", sourcePath, fd.Name(), containerID, err)
+	}
+
+	// after writing to the file descriptor we need to rewind pointer to the beginning of the file before next reading
+	if _, err := fd.Seek(0, os.SEEK_SET); err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return res, fmt.Errorf("could not seek to the beginning of the file %q: %v", fd.Name(), err)
+	}
+
+	if err := tar.ExtractTarStream(destinationPath, fd); err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return res, fmt.Errorf("could not extract runtime artifact %q into the directory %q: %v", sourcePath, destinationPath, err)
+	}
+
+	return utilstatus.NewFailureReason("", ""), nil
+}
+
+// check for new labels and apply to the output image.
+func checkAndGetNewLabels(builder *STI, docker dockerpkg.Docker, tar s2itar.Tar, containerID string) error {
+	glog.V(3).Infof("Checking for new Labels to apply... ")
+	newLabelsFileName := "new_labels.json"
+	sourceFilePath := "/opt/docker/" + newLabelsFileName
+	destinationDir := builder.config.WorkingDir + "/upload/"
+	res, err := downloadAndExtractFileFromContainer(docker, tar, sourceFilePath, destinationDir, containerID)
+	if err != nil {
+		builder.result.BuildInfo.FailureReason = res
+	} else {
+		// check and read the 'newLabelsFileName'.
+		newLablesFilepath := destinationDir + newLabelsFileName
+		fd, ferr := os.Open(newLablesFilepath)
+		if ferr != nil {
+			glog.V(3).Infof("New LABELs file '%s' could not be opened... continuing...", newLablesFilepath)
```

Comment:
this should be a failure.  if they provided a file and we extracted it but then couldn't read it locally, we don't want to commit an image that doesn't have the metadata they supplied.